### ISSUE TITLE
does asdf support virtual datasets?

### ISSUE THREAD

**Issue by ggggggggg** [spacetelescope/asdf#204]
I can't tell from reading the specification what transformations do, so I am asking here. Can I specify a dataset `a` that draws data from dataset `b`, but then applies some transformation to the data before returning it?  For example `b` = [1 2 3], `a`=b+4, so when I read values from `a` I would read [5 6 7].

The whole discussion of transformations seems tailored to the astronomy community, and is hard to follow if you don't know what wcs is.

**Comment by nden** [comment]
@ggggggggg Although the motivation for this was `WCS` the transforms are written to be general, see [asdf-standard](http://asdf-standard.readthedocs.io/en/latest/index.html) for the schemas. The python implementation in `asdf` is based on [astropy.modeling](http://docs.astropy.org/en/stable/modeling/index.html). `ASDF` serializes the transforms to a file.

I am not sure what you mean by `virtual dataset` but perhaps the example below does what you are describing:

```
from asdf import AsdfFile
import numpy as np
from astropy.modeling import models

# write data to a file `target.asdf`
tree = {
    'a': np.arange(0, 10)
}

target = AsdfFile(tree)
target.write_to('target.asdf')

# reference the data in a different file (`source.asdf`)and write a transform.
ff = AsdfFile()
shift = models.Shift(4)
with AsdfFile.open('target.asdf') as target:
    ff.tree['mydata'] = target.make_reference(['a'])
    ff.tree['model'] = shift 
    ff.write_to('source.asdf')

# now read the data from `source.asdf`
f  = AsdfFile.open('source.asdf')
f.resolve_references()
print f.tree['model'](f.tree['mydata'])

```
Does this answer your question?

**Comment by ggggggggg** [comment]
Yes, you answered my question. That is example is very helpful for understanding how one would use a model, thank you for providing it. 

I've been looking at various file formats, and saw the feature I called "virtual datasets"  (they call them derived fields) when looking at the [dirfile format](http://getdata.sourceforge.net/dirfile.html#fields). When I read about asdf I didn't understand what models were, so I thought they were the same thing. They're quire similar. It's basically just a but of extra magic on top of your example so that one doesn't need to load model explicitly. For example you could imagine that the following would result in the two print statements being identical.
```
from asdf import AsdfFile
import numpy as np
from astropy.modeling import models

# write data to a file `target.asdf`
tree = {
    'a': np.arange(0, 10)
}

target = AsdfFile(tree)
target.write_to('target.asdf')

# reference the data in a different file (`source.asdf`)and write a transform.
ff = AsdfFile()
shift = models.Shift(4)
with AsdfFile.open('target.asdf') as target:
    ff.tree['mydata'] = target.make_reference(['a'])
    ff.tree['model'] = shift 
    ff.tree["mydatawithmodel"] = ff.make_reference(["mydata"], ["model"])
    ff.write_to('source.asdf')

# now read the data from `source.asdf`
f  = AsdfFile.open('source.asdf')
f.resolve_references()
print f.tree['model'](f.tree['mydata'])
print f.tree["mydatawithmodel"]
```
